README.md
# set_value

> Create nested values and any intermediaries using list notation `['a','b',c']` OR dot notation `"a.b.c"`.

This is a `Dart` copy of the `js` project [set-value](https://www.npmjs.com/package/set-value) by [Jon Schlinkert](https://github.com/jonschlinkert).
Please consider following either (or both) projects and starring the projects to show your :heart: and support.
Also, feel free to make pull requests, open issues, or just let me know if I've messed up. The goal is to keep this project in-line with the original as far as possible.
Lastly, please consider following the project author [Ferdinand Steenkamp](https://github.com/Ferdzzzzzzzz/).


## Install
Install with [pub.dev](https://pub.dev/), add this to your `pubspec.yaml`:

    dependencies:
        set_value:
Then run `flutter pub get`.

## Usage

Import the package:
    
    import 'package:set_value/set_value.dart';

### Using `dot` notation:
    
    void main() {
        
        var mock = <String, dynamic>{}
        var setValue = SetValue();
        
        // set a value in a map
        mock = setValue.setDot(mock, 'a.b.c', 'Value', splitAt = '.', escapeWith = '\\');
        
        // unset value in a map
        mock = setValue.unsetDot(mock, 'a.b.c');
        
        //get value in a map, returns null if it doesn't exists
        print(setValue.getDot(mock, 'a.b.c'));
    }

Notes:

*Escaping:* You can escape a split character using \\\\, alternatively, you can add your own escape character using the optional `escapeWith` option.

*Dot Notation:* The dot notation oviously uses...dots. Alternatively, you can add your own dot character using the optional `splitAt` option.

### Using `List<String>` method:
    
    void main() {
        
        var mock = <String, dynamic>{}
        var setValue = SetValue();
        
        // set a value in a map
        mock = setValue.set(mock, ['a', 'b', 'c'], 'Value');
        
        // unset value in a map
        mock = setValue.unset(mock, ['a', 'b', 'c']);
        
        //get value in a map, returns null if it doesn't exists
        print(setValue.get(mock, ['a','b','c']));
    }



### About

Pull requests and issues always welcome. There is a lot more information on this project at the original [set-value](https://www.npmjs.com/package/set-value). 

### Author

**Ferdinand Steenkamp**

* [Website](https://ferdzz.com)
* [GitHub Profile](https://github.com/Ferdzzzzzzzz)
* [Twitter Profile](https://twitter.com/therealferdz)
* [LinkedIn Profile](https://www.linkedin.com/in/ferdinand-steenkamp-552335181/)



### License

Copyright © 2020, [Ferdinand Steenkamp](https://github.com/Ferdzzzzzzzz).
Released under the [BSD License](LICENSE).





CHANGELOG.md
## 1.0.4

- Fixed docs

## 1.0.3

- Added get and getDot functions

## 1.0.2

- Added more optional params
- Fixed small bugs
- Added dot notation
- Added tests
- Document fixes

## 1.0.1

- Document fix

## 1.0.0

- Initial version, created by Stagehand


test/set_value_test.dart
import 'package:set_value/set_value.dart';
import 'package:test/test.dart';

void main() {
  group('Test List Notation', () {
    test('Test set function', () {
      var setValue = SetValue();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'}
      };

      testMap =
          setValue.set(testMap, ['Key 4', 'Key 4.1', 'Key 4.1.1'], 'Value 4');
      testMap = setValue.set(testMap, ['Key 5'], [1, 2]);

      var assertTest = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 4': {
          'Key 4.1': {
            'Key 4.1.1': 'Value 4',
          }
        },
        'Key 5': [1, 2]
      };

      expect(testMap, equals(assertTest));
    });

    test('Test unset function', () {
      var setValue = SetValue();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
            'Key 5.2.2': 'Val 5.2.2',
          }
        }
      };

      testMap = setValue.unset(testMap, ['Key 5', 'Key 5.2', 'Key 5.2.2']);
      testMap = setValue.unset(testMap, ['Key 1']);

      var assertTest = <String, dynamic>{
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
          }
        }
      };

      expect(testMap, equals(assertTest));
    });
  });

  group('Test Dot Notation', () {
    test('Test setDot function', () {
      var setValue = SetValue();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'}
      };

      testMap =
          setValue.setDot(testMap, 'Key 4.Key 4\\.1.Key 4\\.1\\.1', 'Value 4');
      testMap = setValue.setDot(testMap, 'Key 5', [1, 2]);

      var assertTest = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 4': {
          'Key 4.1': {
            'Key 4.1.1': 'Value 4',
          }
        },
        'Key 5': [1, 2]
      };

      expect(testMap, equals(assertTest));
    });

    test('Test unsetDot function', () {
      var setValue = SetValue();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
            'Key 5.2.2': 'Val 5.2.2',
          }
        }
      };

      testMap = setValue.unsetDot(testMap, 'Key 5.Key 5\\.2.Key 5\\.2\\.2');
      testMap = setValue.unsetDot(testMap, 'Key 1');

      var assertTest = <String, dynamic>{
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
          }
        }
      };

      expect(testMap, equals(assertTest));
    });
  });
}


lib/src/set_value_base.dart
class SetValue {
  SetValue();

  ///
  /// set
  ///
  /// Takes a json object, path and value,
  /// returns the object with that value set
  /// at the path
  ///
  /// @Map<String,dynamic> target
  /// @List<String> path
  /// @dynamic value
  ///
  Map<String, dynamic> set(
    Map<String, dynamic> target,
    List<String> path,
    dynamic value,
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      _result(target, path[0], value);
      return target;
    }

    var orig = target;
    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target[prop] is Map<String, dynamic>)) {
        target[prop] = <String, dynamic>{};
      }

      if (i == len - 1) {
        _result(target, prop, value);
        break;
      }

      target = target[prop];
    }

    return orig;
  }

  ///
  /// unset
  ///
  /// Takes a json object and unsets a key
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  ///
  Map<String, dynamic> unset(
    Map<String, dynamic> target,
    List<String> path,
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      target.remove(path[0]);
      return target;
    }

    var orig = target;
    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target.containsKey(prop))) {
        break;
      }

      if (i == len - 1) {
        target.remove(path[i]);
        break;
      }

      target = target[prop];
    }

    return orig;
  }

  ///
  /// get
  ///
  /// Takes a json object and path and returns the value at that path
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  ///
  ///
  dynamic get(
    Map<String, dynamic> target,
    List<String> path,
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      return target[path[0]];
    }

    var orig = target;
    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target.containsKey(prop))) {
        break;
      }

      if (i == len - 1) {
        return target[path[i]];
        break;
      }

      target = target[prop];
    }
    return null;
  }

  ///
  /// setDot
  /// Add a value to the Map key using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @dynamic value
  ///
  Map<String, dynamic> setDot(
    Map<String, dynamic> target,
    String path,
    dynamic value, {
    String splitAt = '.',
    String escapeWith = '\\',
  }) {
    return set(target, _split(path, splitAt, escapeWith), value);
  }

  ///
  /// unsetDot
  /// Remove a key from the map using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @optional String splitAt
  /// @optional String escapeWith
  ///
  Map<String, dynamic> unsetDot(
    Map<String, dynamic> target,
    String path, {
    String splitAt = '.',
    String escapeWith = '\\',
  }) {
    return unset(target, _split(path, splitAt, escapeWith));
  }


  ///
  /// getDot
  /// Get a value at any path using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @optional String splitAt
  /// @optional String escapeWith
  ///
  dynamic getDot(
      Map<String, dynamic> target,
      String path, {
        String splitAt = '.',
        String escapeWith = '\\',
      }) {
    return get(target, _split(path, splitAt, escapeWith));
  }

  ///
  /// _result
  /// Add a value to the Map key
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @dynamic value
  ///
  void _result(
    Map<String, dynamic> target,
    String path,
    dynamic value,
  ) {
    target[path] = value;
  }

  ///
  /// _split
  /// Split an String with dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String splitAt
  /// @String escapeWith
  ///

  List<String> _split(String path, String splitAt, String escapeWith) {
    var keys = path.split(splitAt);

    var res = <String>[];
    String prop;

    for (var i = 0; i < keys.length; i++) {
      prop = keys[i];
      var lastChar = prop.length - 1;
      while (prop.substring(lastChar) == escapeWith) {
        prop = prop.substring(0, lastChar) + splitAt + keys[++i];
        lastChar = prop.length - 1;
      }
      res.add(prop);
    }
    return res;
  }
}


example/set_value_example.dart
import 'package:set_value/set_value.dart';
import 'package:pretty_json/pretty_json.dart';

void main() {
  //List Example
  var mock = <String, dynamic>{
    'Key 1': 'Val 1',
    'Key 2': 'Val 2',
    'Key 3': {'Key 3.1': 'Val 3'}
  };

  var setValue = SetValue();

  /*
  * Below we set three different keys on the map,
  * two of which are nested.
  * */

  mock = setValue.set(mock, ['Key 4'], 'Val 4');
  mock = setValue.set(mock, ['Key 5', 'Key 5.1'], 'Val 5.1');

  /*
  * Below we remove Key 2 and it's value from the map
  * */

  mock = setValue.unset(mock, ['Key 2']);

  /*
  * Here we expect an output of:
  * {
  *   'Key 1': 'Val 1',
  *   'Key 3': {
  *     'Key 3.1': 'Val 3'
  *   }
  *   'Key 4': 'Val 4',
  *   'Key 5':  {
  *     'Key 5.1': 'Val 5.1',
  *   }
  * }
  * */

  print(prettyJson(mock));

  //Dot Notation Example
  //reset the Map
  mock = <String, dynamic>{
    'Key 1': 'Val 1',
    'Key 2': 'Val 2',
    'Key 3': {'Key 3.1': 'Val 3'}
  };

  setValue = SetValue();

  /*
  * Below we set three different keys on the map,
  * two of which are nested.
  * */

  mock = setValue.setDot(mock, 'Key 4', 'Val 4');

  //Remember to escape unwanted splits using \\
  //Alternatively, you can specify your own split character using the
  //option 'splitAt' param
  mock = setValue.setDot(mock, 'Key 5.Key 5\\.1', 'Val 5.1');

  /*
  * Below we remove Key 2 and it's value from the map
  * */

  mock = setValue.unsetDot(mock, 'Key 2');

  /*
  * Here we expect an output of:
  * {
  *   'Key 1': 'Val 1',
  *   'Key 3': {
  *     'Key 3.1': 'Val 3'
  *   }
  *   'Key 4': 'Val 4',
  *   'Key 5':  {
  *     'Key 5.1': 'Val 5.1',
  *   }
  * }
  * */

  print(prettyJson(mock));

  print(setValue.get(mock, ['Key 5','Key 5.1']));
  print(setValue.getDot(mock, 'Key 5.Key 5\\.1'));
}
